### PR TITLE
Add extensions block to manifest

### PR DESCRIPTION
**Pull request by punya**
Also, manifest is now output as JSON (which is valid YAML).

Implements https://github.com/palantir/sls-spec/pull/15.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/gradle-java-distribution/143)

### CODE REVIEW

**Comment by uschi2000** [comment]
@nmiyake are you cool with this json/yaml format?

File: src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy [hunk 1]
```diff
@@ -43,18 +43,24 @@ class CreateManifestTask extends BaseTask {
         return String.valueOf(project.version)
     }
 
+    @Input
+    public Map<String, Object> getExtraProperties() {
```
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy]: doesn't have test coverage

**Comment by nmiyake** [comment]
Yup LGTM

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWjTX3x8Ll6LdPGElt_:b-vn0kij)_

**Comment by nmiyake** [comment]
Adding @bmoylan for SA

File: src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy [hunk 2]
```diff
@@ -43,18 +43,24 @@ class CreateManifestTask extends BaseTask {
...
 
     @TaskAction
     void createManifest() {
-        EmitFiles.replaceVars(
-                CreateManifestTask.class.getResourceAsStream('/manifest.yml'),
-                getManifestFile().toPath(),
-                ['@serviceGroup@'  : getServiceGroup(),
-                 '@serviceName@'   : getServiceName(),
-                 '@serviceVersion@': getProjectVersion()])
+        getManifestFile().setText(JsonOutput.prettyPrint(JsonOutput.toJson(getExtraProperties() + [
```
Comment (uschi2000) [inline src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy]: do we want to group the extra properties under a single key such as `ext`, `extensions`, `extras` or similar?

**Review by uschi2000 — CHANGES_REQUESTED** [review]

**Comment by punya** [comment]
I can make it YAML again if desired -- I'll just have to add SnakeYAML or something similar to the dependencies. Let me know.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy, line 47 at r1](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-KWjN17YnQVaZ3n0Nh8E-r1-47:-KWjh6kdxmWtqepvdycS:b9e2xig) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/bf31ac4855db1b9bf84f61c7db0d9e3f4aacc4f7/src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy#L47)):_
<details><summary>

<i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>



> doesn't have test coverage
> </details>
> Will add in the next round. I wanted feedback on the overall approach before doing that.

---

_[src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy, line 58 at r1](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-KWjN17YnQVaZ3n0Nh8E-r1-58:-KWjhCDZ279xJYP39wE3:b-i1rzrk) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/bf31ac4855db1b9bf84f61c7db0d9e3f4aacc4f7/src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy#L58)):_
<details><summary>

<i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>



> do we want to group the extra properties under a single key such as `ext`, `extensions`, `extras` or similar?
> </details>
> The spec says you can add properties willy-nilly at the top level, but there was discussion (https://github.com/palantir/sls-spec/issues/13) of using a top-level extensions key. Since that discussion seemed to have stalled out, I went with the current version of the spec. I'd be happy to switch this over to a top-level key once the spec PR merges, but for now I don't want to block this PR on that debate.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWjgyy8vqENeI3rTrEG:bi2uv7z)_

**Comment by punya** [comment]
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

_[src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy, line 47 at r1](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-KWjN17YnQVaZ3n0Nh8E-r1-47:-KWk7BD7aaTIOD-x8Inp:b-896fix) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/bf31ac4855db1b9bf84f61c7db0d9e3f4aacc4f7/src/main/groovy/com/palantir/gradle/javadist/tasks/CreateManifestTask.groovy#L47)):_
<details><summary>

<i>Previously, punya (Punyashloka Biswal) wrote…</i></summary>



> Will add in the next round. I wanted feedback on the overall approach before doing that.
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143)_

**Comment by uschi2000** [comment]
Punya, check the sls spec changes please; we want those things to show up under a "extensions" key.

---

Reviewed 1 of 3 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWp2DpPYAW4iM4UL-jz:b-67lh2d)_

**Comment by uschi2000** [comment]
readme needs adjustment.

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[src/main/groovy/com/palantir/gradle/javadist/DistributionExtension.groovy, line 116 at r3](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-KWsHrl3SHJQh9d2mkFX:-KWsHrl3SHJQh9d2mkFY:blus95s) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/ea61e2fabc6ed82a2b70ef2a529e524e381744b9/src/main/groovy/com/palantir/gradle/javadist/DistributionExtension.groovy#L116)):_

> ``` Groovy
>     }
> 
>     public void setExtensionsBlock(Map<String, Object> extensionsBlock) {
> ```

can you follow the same pattern as with `env` and also add some tests to DistributionExtensionTest?

---

_[src/main/groovy/com/palantir/gradle/javadist/DistributionExtension.groovy, line 116 at r3](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-KWsNSSs78PxOTP8SQ_Y:-KWsNSSs78PxOTP8SQ_Z:b-b3h0i3) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/ea61e2fabc6ed82a2b70ef2a529e524e381744b9/src/main/groovy/com/palantir/gradle/javadist/DistributionExtension.groovy#L116)):_

> ``` Groovy
>     }
> 
>     public void setExtensionsBlock(Map<String, Object> extensionsBlock) {
> ```

I wonder what the right name is for this. maybe `manifestExtension`?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWsNJhrEic3021nv5Pf:b-1lcdjj)_

**Comment by uschi2000** [comment]
:lgtm:
@markelliot want to take a look?

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWsSNbhMMB3atcBW-nQ:btkjqf4)_

**Review by markelliot — APPROVED** [review]

**Comment by punya** [comment]
Just realized that I still haven't updated the Readme, will do that shortly.

**Comment by punya** [comment]
@uschi2000 ready to merge, I think!

**Comment by uschi2000** [comment]
:lgtm:

---

Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143#-:-KWt9BFhjaMnbJfLfBLb:bnfp4nl)_

**Comment by uschi2000** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/143)_

**Review by uschi2000 — APPROVED** [review]

**Comment by punya** [comment]
@uschi2000 @markelliot @nmiyake when will this get released?

**Comment by markelliot** [comment]
I think this went in with some of the TLS changes and we're looking to do an RC to test some things. But Rob and Joe are out this week, so think it's going to be next week.

**Comment by bmoylan** [comment]
Would be great to have a release this week, I need to add extensions for something shipping on ~Friday and would prefer to not hack them in or depend on a snapshot

**Pull request merged**